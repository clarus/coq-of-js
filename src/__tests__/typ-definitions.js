// @flow
import {compileAndPrint} from "../compiler/index.js";

describe("type synonyms", () => {
  it("handles type synonyms of constants", () => {
    expect(
      compileAndPrint(
        `type b = boolean;
type s = string;
type v = void;
type n = null;
type f = number;
type o = {};
type e = empty;
`,
      ),
    ).toMatchSnapshot();
  });
});

describe("enums", () => {
  it("handles type definition of enums", () => {
    expect(
      compileAndPrint(`type SayHi = "hi" | 'hello' | "h";`),
    ).toMatchSnapshot();

    expect(compileAndPrint('type Single = "s";')).toMatchSnapshot();
  });

  it("does not handle other elements than strings", () => {
    expect(
      compileAndPrint(`type SayHi = "hi" | 'hello' | 12;`),
    ).toMatchSnapshot();
  });

  it("does not handle other literals than string", () => {
    expect(compileAndPrint(`type Boo = false;`)).toMatchSnapshot();

    expect(compileAndPrint(`type Num = 12;`)).toMatchSnapshot();
  });
});

describe("records", () => {
  it("does handle records", () => {
    expect(
      compileAndPrint(`type Status = {
  message: string,
  quantity: number,
};
`),
    ).toMatchSnapshot();
  });

  it("does not handle spreads", () => {
    expect(
      compileAndPrint(`type Status = {
  message: string,
  quantity: number,
  ...A,
};
`),
    ).toMatchSnapshot();
  });
});

describe("sum types", () => {
  it("handles sum types", () => {
    expect(
      compileAndPrint(`type Status =
  | {
      type: "Error",
      message: string,
    }
  | {
      type: "Loading",
    }
  | {
      type: "Nothing",
    }
  | {
      type: "Success",
      fresh: boolean,
      "value": string,
    };
`),
    ).toMatchSnapshot();

    expect(
      compileAndPrint(`type Status =
  | {
      type: "Error",
      message: string,
    };
`),
    ).toMatchSnapshot();
  });

  it("shows errors for sum types", () => {
    expect(
      compileAndPrint(`type Status =
  | {
      message: string,
    }
  | {
      type: "Loading",
    };
`),
    ).toMatchSnapshot();

    expect(
      compileAndPrint(`type Status =
  | {
      type: "Error",
      message: string,
    }
  | {
      type: number,
    };
`),
    ).toMatchSnapshot();

    expect(
      compileAndPrint(`type Status =
  | {
    type: "Spread",
      ...A,
    };
`),
    ).toMatchSnapshot();

    expect(
      compileAndPrint(`type Status =
  | {
    type: "Spread",
    }
  | string;
`),
    ).toMatchSnapshot();

    expect(compileAndPrint(`type Status = number | string;`)).toMatchSnapshot();
  });
});
