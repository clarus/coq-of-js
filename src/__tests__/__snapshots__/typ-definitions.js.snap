// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does handle records 1`] = `
"Module Status.
  Record t := {
    message : string;
    quantity : Z;
  }.
End Status.
Definition Status := Status.t."
`;

exports[`does not handle enums with other elements than strings 1`] = `
"> 1 | type SayHi = \\"hi\\" | 'hello' | 12;
    |                              ^^

Only strings are handled in enums"
`;

exports[`does not handle other literals than string 1`] = `
"> 1 | type Boo = false;
    |           ^^^^^

Boolean literals in types are not handled"
`;

exports[`does not handle other literals than string 2`] = `
"> 1 | type Num = 12;
    |           ^^

Number literals in types are not handled"
`;

exports[`does not handle records with spreads 1`] = `
"  2 |       message: string,
  3 |       quantity: number,
> 4 |       ...A,
    |      ^^^^
  5 |     };

Expected named property"
`;

exports[`handles sum types 1`] = `
"Module Status.
  Module Error.
    Record t := {
      message : string;
    }.
  End Error.

  Module Success.
    Record t := {
      fresh : bool;
      value : string;
    }.
  End Success.

  Inductive t :=
  | Error (_ : Error.t)
  | Loading
  | Nothing
  | Success (_ : Success.t).
End Status.
Definition Status := Status.t."
`;

exports[`handles sum types 2`] = `
"Module Status.
  Module Error.
    Record t := {
      message : string;
    }.
  End Error.

  Inductive t :=
  | Error (_ : Error.t).
End Status.
Definition Status := Status.t."
`;

exports[`handles type definition of enums 1`] = `
"Module SayHi.
  Inductive t :=
  | hi
  | hello
  | h.
End SayHi.
Definition SayHi := SayHi.t."
`;

exports[`handles type definition of enums 2`] = `
"Module Single.
  Inductive t :=
  | s.
End Single.
Definition Single := Single.t."
`;

exports[`handles type synonyms of constants 1`] = `
"Definition b : Type := bool.

Definition s : Type := string.

Definition v : Type := unit.

Definition n : Type := unit.

Definition f : Type := Z.

Definition o : Type := unit.

Definition e : Type := Empty_set."
`;

exports[`shows errors for sum types 1`] = `
"  1 | type Status =
> 2 |   | {
    |    ^
> 3 |       message: string,
    | ^^^^^^^^^^^^^^^^^^^^^^
> 4 |     }
    | ^^^^^
  5 |   | {
  6 |       type: \\"Loading\\",
  7 |     };

Expected at least one field with the name \`type\`"
`;

exports[`shows errors for sum types 2`] = `
"  5 |     }
  6 |   | {
> 7 |       type: number,
    |            ^^^^^^
  8 |     };

Expected a string literal"
`;

exports[`shows errors for sum types 3`] = `
"  2 |   | {
  3 |      type: \\"Spread\\",
> 4 |       ...A,
    |      ^^^^
  5 |     };

Expected a named property"
`;

exports[`shows errors for sum types 4`] = `
"  3 |      type: \\"Spread\\",
  4 |     }
> 5 |   | string;
    |    ^^^^^^

Only objects are handled in sum types"
`;

exports[`shows errors for sum types 5`] = `
"> 1 | type Status = number | string;
    |              ^^^^^^^^^^^^^^^

Only handle unions of strings or objects with a \`type\` field"
`;
